### ISSUE TITLE
Automatic font outlines are too thin in higher resolution (800x600 and up)

### ISSUE THREAD

**Issue by fernewelten** [adventuregamestudio/ags#878]
This is an issue I reported over a year ago, in the AGS forums bug tracker (not here in github). As far as I can see, it's still open:

Automatic font outlines in LucasArts style have a fixed width of 1 pixel for ttf fonts and 0.6 pixel (I think) for SCI fonts.This is fine for 320x200 games or 640x400 games, but at higher resolutions the texts aren't offset enough from the background and so become **very hard** to read. 

This is so off-putting that I've read of programmers that refuse to use LucasArts mode for higher resolutions just because of this one issue, on the AGS boards.

So, when the option "automatic outline" is selected for some font, it should be possible to also select, in pixels, **how wide** the outline should be. This could be implemented in a fully backward-compatible way if the outline width "0" would select the old calculation and be the default unless changed.

To implement this, the following needs to be done:
1. Logic must be provided that calculates outlines of variable thickness at runtime. One straight-forward way to do it is to only change EIGHT code lines (where `outlineDist` is the outline thickness, already declared in the surrounding code), namely:
In file ags/Engine/ac/display.cpp, around line ~450, from: 
```
wouttextxy(ds, xxp - outlineDist, yyp, usingfont, outline_color, texx);
wouttextxy(ds, xxp + outlineDist, yyp, usingfont, outline_color, texx);
wouttextxy(ds, xxp, yyp + outlineDist, usingfont, outline_color, texx);
wouttextxy(ds, xxp, yyp - outlineDist, usingfont, outline_color, texx);
wouttextxy(ds, xxp - outlineDist, yyp - outlineDist, usingfont, outline_color, texx);
wouttextxy(ds, xxp - outlineDist, yyp + outlineDist, usingfont, outline_color, texx);
wouttextxy(ds, xxp + outlineDist, yyp + outlineDist, usingfont, outline_color, texx);
wouttextxy(ds, xxp + outlineDist, yyp - outlineDist, usingfont, outline_color, texx);
```
to:
```
int const max_sq_euclid_dist =
            outlineDist * outlineDist + outlineDist ;
for (int xdiff = -outlineDist; xdiff <= outlineDist; xdiff++)
    for (int ydiff = -outlineDist; ydiff <= outlineDist; ydiff++)
    {
        if (xdiff * xdiff + ydiff * ydiff > max_sq_euclid_dist)
            continue;
        if (0 == xdiff && 0 == ydiff)
            continue;
        wouttextxy(ds, xxp + xdiff, yyp + ydiff, usingfont, outline_color, texx);
    }
```
That's it. 

2. For each font, the outline thickness must be stored. This will probably entail extending the `struct FontInfo` in ags/Common/font, adding the line marked with 3 asterics as follows:

```
struct FontInfo
{
    // General font's loading and rendering flags
    unsigned char Flags;
    // Font size, in points (basically means pixels in AGS)
    int           SizePt;
    // Outlining font index, or auto-outline flag
    char          Outline;
    char OutlineDist; // ***
    // Custom vertical render offset, used mainly for fixing broken fonts
    int           YOffset;
    // custom line spacing between two lines of text (0 = use font height)
    int           LineSpacing;

    FontInfo();
};
```

3. The game file must be extended in such a way that the modified `FontInfo` structure is loaded and saved; I don't know yet what that entails.

4. The editor must be extended in such a way that 
- for every font, the field ```OutlineDist``` becomes visible whenever OutlineStyle is set to "Automatic";
- the default for ```OutlineDist``` is 0 unless set to another value
I don't know C sharp at all, and I haven't fount out yet where the magic of the Editor even happens.

5. The code in In file ags/Engine/ac/display.cpp must be extended in such a way that the value set in the editor is read in. Something along the line:
```
    int const ROV = ReadInOutlineValueForThisFont();
    if (ROV > 0) outlineDiust = ROV;
```

Any pointers that can put me up to speed about 3. to 5. above?

**Comment by ivan-mogilko** [comment]
I could swear I've seen similar request on github recently (I thought it was @selmiak), but could not find it so maybe it's a false memory and we discussed this somewhere else.

Anyway, for starters, I only wanted to say that the above outline drawing may not be the optimal solution, because engine may end up drawing same line(s) of text many dozen of times. This may work as a starting stage, but imho someone should look into implementing outline support per each FontRenderer implementation. For instance, I may be wrong, but I believe seeing a reference to outline support in alfont library (for TTF fonts). Bitmap fonts renderer may have its own optimised way to draw pixels around the glyphs.

Besides perfomance issue, something we briefly discussed before, spreading outline like this in all directions may result in blank spaces in letters filled with outline too (think of letters like 'A' for example). Honestly IDK if that's the desired effect. Of course "automatic outline" means that user handles visual control to the engine, so maybe it's okay.

Putting that aside for now, regarding the suggested algorithm, wouldn't it be better to have multiple straightforward loops instead? That may be bit more lines, but less checks (they only return positive in tiny number of cases) and possibly executing faster, e.g.:
<pre>
for (int xdiff = -outlineDist; xdiff < 0; xdiff++)
{
    for (int ydiff = -outlineDist; ydiff < 0; ydiff++)
        wouttextxy(ds, xxp + xdiff, yyp + ydiff, usingfont, outline_color, texx);
    for (int ydiff = 1; ydiff <= outlineDist; ydiff++)
        wouttextxy(ds, xxp + xdiff, yyp + ydiff, usingfont, outline_color, texx);
}
for (int xdiff = 0; xdiff <= outlineDist; xdiff++)
{
    for (int ydiff = -outlineDist; ydiff < 0; ydiff++)
        wouttextxy(ds, xxp + xdiff, yyp + ydiff, usingfont, outline_color, texx);
    for (int ydiff = 1; ydiff <= outlineDist; ydiff++)
        wouttextxy(ds, xxp + xdiff, yyp + ydiff, usingfont, outline_color, texx);
}
</pre>
This may be tidied further by incrementing actual coordinate rather than keeping summing up xxp+xdiff.
just a thought.


I will reply to other points in a short while.

**Comment by AlanDrake** [comment]
It might be worthwhile to implement a cache system for the font outlines.

**Comment by fernewelten** [comment]
> spreading outline like this in all directions may result in blank spaces in letters filled with outline too

Yes, the same happens with the code as it is now if the hole in the A has a thickness of 2 or less. 

I remember having digged into outlines for .ttf fonts. Trouble with that is, the term "outline" has a formal definition in the .ttf context and that is different from what is meant here. IIRC they mean that the strokes that make up the letters are hollow, e.g.,, the letter "l" is an upright rectangle you can see through and the letter "o" is two slim concentric circles.

In principle, that could be used to draw an outline around another font that is non-hollow. But the ttf standard does so many wonderful things with kerning and spacing and such that in practice, you can never get the two fonts to align properly, one over the other, for longer texts. That really does work with SCI fonts, but I've never seen this work with .ttf fonts.

**Comment by fernewelten** [comment]
> because engine may end up drawing same line(s) of text many dozen of times.

Specifically, (2t + 1)² - 1 times where t is the thickness, or 4t² + 2t times. For thicknesses of at least 5 or so this would probably be often enough to justify @AlanDrake 's idea of caching the outline bitmap.

**Comment by messengerbag** [comment]
If you use a buffer you just need 2(2t+1) drawing operations: First you draw the horizontal offsets of the text onto the buffer, then you draw vertical offsets of that buffer to form the outline (or vice versa).

That is assuming you want a square outline. With thick outlines, a circular or rounded profile may be better. This can also be optimized similarly, but is slightly more complicated. My [SpeechBubble module](https://github.com/messengerbag/SpeechBubble/blob/master/SpeechBubble.asc) has AGS code for either case (starting from line 883 for rounded outlines, and from line 910 for square).

Antialiased text presents serious challenges for any outline method based on a bitmap raster. Most seriously, details that are not fully opaque (e.g. strokes thinner than 1 px) will produce a semi-transparent outline. Therefore, for TrueType fonts outlining should ideally be done in the font library, by stroking/inflating/expanding the outline of each vector glyph.

**Comment by fernewelten** [comment]
Depending how low-level you go, it should be possible to convert any partially opaque pixel in a bitmap to a completely opaque pixel whilst leaving all the completely transparent pixels unmodified with reasonable ease.
On a opacity scale of 0 == transparent to 255 == opaque:
- add 254 to the alpha component of each pixel (1 atomic bitmap operation)
- subtract 254 from the alpha component of each pixel (1 atomic operation)
- multiply the alpha component of each pixel by 255 (1 atomic operation).

**Comment by ivan-mogilko** [comment]
Regarding the implementation (points 1-5 mentioned in the original post):

> 1. Logic must be provided ...

The code location is right of course, from the looks of the suggested code you'd like to have a rounded outline.
My question is, do you think it may worth to have another option for the outline style there (e.g. switch between rounded and square outline)?

---

> 2. For each font, the outline thickness must be stored ...

May I suggest to call new variable OutlineWidth or Thick(ness) instead of "dist"? That seem more clear in my opinion.

---

> 3. The game file must be extended in such a way that the modified FontInfo structure is loaded and saved; I don't know yet what that entails.
> 5. The code in In file ags/Engine/ac/display.cpp must be extended in such a way that the value set in the editor is read in.

The engine's side is pretty trivial actually.

1) you need to increment the game format index, see here: https://github.com/adventuregamestudio/ags/blob/ags3/Common/ac/game_version.h#L137

2) FontInfo structs are loaded in GameSetupStruct::read_font_infos(), see here:
https://github.com/adventuregamestudio/ags/blob/ags3/Common/ac/gamesetupstruct.cpp#L118
You may notice that there are already branches for pre-3.5.0 and 3.5.0 data formats. Reading of a new data may be appended after 3.5.0 data is read with similar format index check.

3) To keep compatible with earlier versions that don't have this parameter in data you may simply assign new variable in each FontInfo to default value of 1.
This may be done either in FontInfo's constructor, read_font_infos() function, or UpgradeFonts() function (see https://github.com/adventuregamestudio/ags/blob/ags3/Common/game/main_game_file.cpp#L470).
Latter function is mostly meant for converting and filling data which is only possible after it was read from older format, so in your case FontInfo's constructor may actually be best option (since thickness of 1 is a good default anyway).

4) You do not need to call any extra "reading" during the drawing itself, just use previously set variable like -
<pre>
outlineDist = game.fonts[usingfont].OutlineDist; // or whatever you call it
</pre>

If you need to differentiate between old algorithm and new one, which has outline thickness explicitly set per font, you just check `loaded_game_file_version`, for example:
<pre>
if (loaded_game_file_version >= kGameVersion_XXX)
{
    do new stuff
}
else
{
    do old stuff
}
</pre>
you may see examples of similar switches in many places in the engine.

---

> 4. The editor must be extended in such a way that
> - for every font, the field OutlineDist becomes visible whenever OutlineStyle is set to "Automatic";
> - the default for OutlineDist is 0 unless set to another value

Now, for the editor part...

1) You add new property to the AGS.Types.Font class here:
https://github.com/adventuregamestudio/ags/blob/ags3/Editor/AGS.Types/Font.cs

Note that AGS Editor has a system set, where each C# property of a game object that is *public*, *not read-only* and not tagged as *"Browsable(false)"* will be serialized automatically in the project file.
This means that simply adding a property, for example, OutlineThickness, with default value of 1 will be just enough and should probably cover backwards compatibility (might need to think this over again later).

2) How do you hide OutlineThickness if OutlineStyle is not automatic?
This is a bit hacky. Currently this is done by making particular object class implement .NET interface called ICustomTypeDescriptor. I think it's better to see existing example, such as Character class:
https://github.com/adventuregamestudio/ags/blob/ags3/Editor/AGS.Types/Character.cs
Notice it inherits ICustomTypeDescriptor and implements number of functions between these two lines:
<pre>
#region ICustomTypeDescriptor Members
...
#endregion
</pre>
Most of these are simply stubs, but the important function that you have to override is ` public PropertyDescriptorCollection GetProperties(Attribute[] attributes)`
What it does it puts all existing type properties in the list, but skips particular ones under certain conditions.
You will have to do similar thing for Font class. With the code like:
<pre>
if ((_outlineStyle != FontOutlineStyle.Automatic) && (property.Name == "OutlineThickness"))
{
    wantThisProperty = false;
}
</pre>
By the way, you could also hide OutlineFont property while you are there, unless style is UseOutlineFont.

3) Finally, game data writing is done by DataFileWriter class:
https://github.com/adventuregamestudio/ags/blob/ags3/Editor/AGS.Editor/DataFileWriter.cs#L1394
Of course you must make sure the way its written and read (by engine) comply to each other.

**Comment by fernewelten** [comment]
Hi Ivan,
Thanks for the **very thorough** explanation!

> **Drawing code**
> have another option for the outline style there (e.g. switch between rounded and square outline)?

I think this should be easy to offer, so why not. The code that I quoted was a special one-time case for a special occasion: I used it to make my recent game have an outline thickness of 3 px. If we want to do this properly and offer it for everyone, then it should be suitably generalized.

> **New FontInfo member**
> call new FontInfo's variable OutlineWidth or Thick(ness) instead of "dist"

I agree, that's sensible.

**Comment by fernewelten** [comment]
So I'll start to have a go at this.

**Comment by fernewelten** [comment]
The colour of the outline is read from the gamestate and saved into the gamestate, as variable `speech_text_shadow` (declared in gamestate.h~line77). However, I haven't seen any way to set it by program or in the Editor to any specific value. So it has to stay its original value the whole time. Am I overlooking something?

Should we offer a way to set this outline colour?

Would it make sense to set (override) this outline colour by character, too, so that e.g. the normal outline character is black, but some specific characters use another color such as red as a special effect? For instance, a devil might have black text on fiery red outline.

**Comment by ivan-mogilko** [comment]
The half of the "play" struct is exposed to script API as "GameState game" struct. https://github.com/adventuregamestudio/ags/blob/ags3/Editor/AGS.Editor/Resources/agsdefns.sh#L2613
In the engine there's a comment that tells which part of the internal struct is actually exported:
https://github.com/adventuregamestudio/ags/blob/ags3/Engine/ac/gamestate.h#L144

In the past we were gradually moving towards replacing raw variables with properties (attributes in AGS script), and in theory same could be done with this outline variable: it may be deprecated and replaced with a property in e.g. struct Game. But that's optional.

> Would it make sense to set (override) this outline colour by character,

This could be added to character's API of course if you think it's worth it, although if I may I'd suggest to separate this into its own task.

**Comment by ivan-mogilko** [comment]
PS. Also, in the past I was writing few articles in repository's wiki:
https://github.com/adventuregamestudio/ags/wiki
Noteably this page on working with Script API: 
https://github.com/adventuregamestudio/ags/wiki/A-quick-guide-to-Script-API

Please suggest if there's an information you think should be added there to make it easier making similar changes to the engine.

**Comment by ericoporto** [comment]
Hey, since you guys are doing this, one cool thing from SpeechBubble is that it let's the outline color be any, so one can make it be a mix from the main font color and some other color or black. Having colored outline is pretty cool. :)

**Comment by fernewelten** [comment]
>… it lets the outline color be any[thing] so one can make it be a mix from the main font color and some other color or black. Having colored outline is pretty cool. :)

Hi Erico, have a look at `game.text_shadow_color` (for documentation, open the in-studio documentation by hitting F1 and click in the index on the heading "game. variables"). If you want to have, e.g., a devil that says his words with a fiery red outline, then do something like
```
int normal_outline = game.text_shadow_color;
game.text_shadow_color = 12; // red
cDevil.Say("Hoooo! I'm a dangerous devil, I am!");
game.text_shadow_color = normal_outline;
Display("Back to normal outlines");
```
Unfortunately, that won't help you for text that is generated automatically such as the text in the status line.

**Comment by ivan-mogilko** [comment]
We've also briefly discussed the possibility to add per-character outline color property couple of comments above.

**Comment by ivan-mogilko** [comment]
@fernewelten , I did some tests, and this is what I found in regards to behavior:

1. Project conversion from old format *editor).

* Something I forgot to mention earlier, you also need to increase project format index, after adding new properties.
  It's defined by constant `LATEST_XML_VERSION_INDEX`.
* Is there a reason to have "OutlineThickness" default value = 0 instead of 1? Will it have special meaning rather than being just a extra case when it has to be converted to 1 later?
  Note that right now engine assumes that thickness = 0 indicates old font format and not only changes it to 1, but also changes style to Squared (there's another problem that I mention below).
* There has to be an explicit conversion from old format done in AGS.Editor.Tasks.SetDefaultValuesForNewFeatures. For previous format, all fonts with Auto outline should be set to style = Square thickness = 1.

2. Game data conversion (engine).

* Data conversion in UpgradeFonts currently has incorrect condition. It is applied only if game format is *3.5.0 or higher*, while logically it should be applied if format is *3.5.0 or lower*. Also, in this case there is no need to test for Thickness = 0, because in these old formats thickness was never a thing.
OTOH, if you mean to correct even new format, adjusting thickness = 0 to 1, then this should be better done separately; also in such case it should not force style to Squared like it does now.
I guess the question here is if thickness = 0 should be a valid value in a new format at all (this is also mentioned above).

3. New functionality (engine).

I found that if I set large thickness, then the engine cannot correctly calculate size of bitmap necessary to store the text line.
In my test this happened when I took font with height = 9 and assigned outline thickness = 9.

**Comment by ericoporto** [comment]
One thing was discussed at some point, would be creating the outlines either at bootup, similar to a non automatic font outline, or cache as needed, as to avoid repeatedly drawing to create the outline effect.

**Comment by ivan-mogilko** [comment]
I believe that may be added separately, as further optimization.

**Comment by ivan-mogilko** [comment]
Was implemented in master branch by #892 / #1018, possibly will be included in 3.5.1 release.
